ITensorGaussianMPS/src/gmps.jl

```diff
@@ -111,52 +136,110 @@ is_annihilation_operator(::OpName"c") = true
     coefs[n] = coef
+    quads[n] = quadrant(term)
     length(term) ≠ 2 && error("Must create hopping Hamiltonian from quadratic Hamiltonian")
-    @assert is_creation_operator(term[1])
-    @assert is_annihilation_operator(term[2])
     sites[n] = ntuple(n -> ITensors.site(term[n]), Val(2))
     nsites = max(nsites, maximum(sites[n]))
   end
-  ElT = all(isreal(coefs)) ? Float64 : ComplexF64
-  h = zeros(ElT, nsites, nsites)
+  # detect coefficient type 
+  coef_type=mapreduce(typeof,promote_type,coefs)
+  ElT = isreal(coefs) ? real(coef_type) : coef_type
+  # fill Hamiltonian matrix with elements
+  h = zeros(ElT, 2 * nsites, 2 * nsites)
+  other_quad = i -> i == 2 ? 1 : 2
   for n in 1:nterms
-    h[sites[n]...] = coefs[n]
+    quad = quads[n]
+    offsets = nsites .* (quad .- 1)
+    h[(sites[n] .+ offsets)...] += coefs[n]
+    other_offsets = nsites .* (other_quad.(quad) .- 1)
+    h[(sites[n] .+ other_offsets)...] += -conj(coefs[n])
   end
-  return h
+  return Hermitian(interleave(h))
 end
 
-# Make a combined hopping Hamiltonian for spin up and down
-function hopping_hamiltonian(os_up::OpSum, os_dn::OpSum)
-  h_up = hopping_hamiltonian(os_up)
-  h_dn = hopping_hamiltonian(os_dn)
+function quadratic_hamiltonian(os_up::OpSum, os_dn::OpSum)
+  h_up = quadratic_hamiltonian(os_up)
+  h_dn = quadratic_hamiltonian(os_dn)
   @assert size(h_up) == size(h_dn)
   N = size(h_up, 1)
-  ElT = promote_type(eltype(h_up), eltype(h_dn))
-  h = zeros(ElT, 2 * N, 2 * N)
-  for i in 1:(2 * N), j in 1:(2 * N)
-    if isodd(i) && isodd(j)
-      i_up, j_up = (i + 1) ÷ 2, (j + 1) ÷ 2
-      h[i, j] = h_up[i_up, j_up]
-    elseif iseven(i) && iseven(j)
-      i_dn, j_dn = i ÷ 2, j ÷ 2
-      h[i, j] = h_dn[i_dn, j_dn]
-    end
+  h = zeros(eltype(h_up), (2 * N, 2 * N))
+  n = div(N, 2)
+  # interlace the blocks of both quadratic hamiltonians
+  h_up = reverse_interleave(Matrix(h_up))
+  h_dn = reverse_interleave(Matrix(h_dn))
+  # super-quadrant (1,1)
+  h[1:2:N, 1:2:N] = h_up[1:n, 1:n]
+  h[2:2:N, 2:2:N] = h_dn[1:n, 1:n]
+  # super-quadrant (2,1)
+  h[(N + 1):2:(2 * N), 1:2:N] = h_up[(n + 1):(2 * n), 1:n]
+  h[(N + 2):2:(2 * N), 2:2:N] = h_dn[(n + 1):(2 * n), 1:n]
+  # super-quadrant (2,2)
+  h[(N + 1):2:(2 * N), (N + 1):2:(2 * N)] = h_up[(n + 1):N, (n + 1):N]
+  h[(N + 2):2:(2 * N), (N + 2):2:(2 * N)] = h_dn[(n + 1):N, (n + 1):N]
+  # super-quadrant (1,2)
+  h[1:2:N, (N + 1):2:(2 * N)] = h_up[1:n, (n + 1):(2 * n)]
+  h[2:2:N, (N + 2):2:(2 * N)] = h_dn[1:n, (n + 1):(2 * n)]
+
+  #convert from blocked to interlaced format. Odd base-rows are spin-up, even are spin-down.
+  return Hermitian(interleave(h))
+end
+
+function hopping_hamiltonian(os::OpSum; drop_pairing_terms_tol=0.0)
+  # convert to blocked format
+  h = reverse_interleave(Matrix(quadratic_hamiltonian(os)))
+  # check that offdiagonal blocks are 0
+  N = div(size(h, 1), 2)
+  tol=max(drop_pairing_terms_tol,eps(real(eltype(h))))
```

Comment:
Instead of this, shouldn't we make the default value equivalent to `drop_pairing_terms_tol=eps(real(eltype(h)))`? I think we should let people set the tolerance to zero if they want, but pick a reasonable default value.

We will have to write a function that gets the equivalent of `eltype(h)` based on `os`. We could define a function `coefficient_type(os::OpSum)`, for example:
```julia
function coefficient_type(os::OpSum)
  return mapreduce(typeof ∘ coefficient, promote_type, terms(os))
end
```
or something like that. Then we can set the default `drop_pairing_terms_tol=eps(real(coefficient_type(os)))`.

Otherwise, we could set the default to `drop_pairing_terms_tol=nothing` and then have a line:
```julia
if isnothing(drop_pairing_terms_tol)
  drop_pairing_terms_tol = eps(real(eltype(h)))
end
```
Maybe that makes the most sense for efficiency (i.e. not having to loop over the coefficients multiple times).

Reply:
I opted for the latter, since it's simpler.